src/handler/mod.rs

```diff
@@ -1265,4 +1507,159 @@ impl Handler {
             .ban_nodes
             .retain(|_, time| time.is_none() || Some(Instant::now()) < *time);
     }
+
+    async fn new_connection(
+        &mut self,
+        enr: Enr,
+        socket_addr: SocketAddr,
+        conn_dir: ConnectionDirection,
+    ) {
+        if let Err(e) = self
+            .service_send
+            .send(HandlerOut::Established(enr, socket_addr, conn_dir))
+            .await
+        {
+            warn!(
+                "Failed to inform of established connection {}, {}",
+                conn_dir, e
+            )
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl HolePunchNat for Handler {
+    async fn on_request_time_out<P: ProtocolIdentity>(
+        &mut self,
+        relay: NodeAddress,
+        local_enr: Enr, // initiator-enr
+        timed_out_nonce: MessageNonce,
+        target_node_address: NodeAddress,
+    ) -> Result<(), NatHolePunchError> {
+        // Another hole punch process with this target may have just completed.
+        if self.sessions.cache.get(&target_node_address).is_some() {
+            return Ok(());
+        }
```

Comment:
Since we return `Ok` here, [the request is reinserted into `self.active_requests`](https://github.com/emhane/discv5/blob/41f0ef64dbf6fa24745787d8e5ad889132d93ba5/src/handler/mod.rs#L542), but the request isn't actually sent. Perhaps we should return `Err` instead of `Ok`?